expr/func_movinwindow_test.go

```diff
@@ -1 +1,357 @@
+		[]models.Series{
+			{
+				Target:     "movingAverage(first,\"10min\")",
+				QueryPatt:  "movingAverage(first,\"10min\")",
+				QueryFrom:  10 + offset10mins,
+				QueryTo:    80,
+				Datapoints: getCopy(avgALast10mins),
+			},
+			{
+				Target:     "movingAverage(second,\"10min\")",
+				QueryPatt:  "movingAverage(second,\"10min\")",
+				QueryFrom:  10 + offset10mins,
+				QueryTo:    80,
+				Datapoints: getCopy(avgALast10mins),
+			},
+		},
+		offset10mins,
+		"10min",
+		"average",
+		1,
+		t)
+
+	testMovingWindow(
+		"Calling movingSum",
+		[]models.Series{
+			{
+				Target:     "first",
+				QueryPatt:  "first",
+				QueryFrom:  10 + offset10mins,
+				QueryTo:    80,
+				Datapoints: getCopy(seriesA),
+			},
+			{
+				Target:     "second",
+				QueryPatt:  "second",
+				QueryFrom:  10 + offset10mins,
+				QueryTo:    80,
+				Datapoints: getCopy(seriesA),
+			},
+		},
+		[]models.Series{
+			{
+				Target:     "movingSum(first,\"10min\")",
+				QueryPatt:  "movingSum(first,\"10min\")",
+				QueryFrom:  10 + offset10mins,
+				QueryTo:    80,
+				Datapoints: getCopy(sumALast10mins),
+			},
+			{
+				Target:     "movingSum(second,\"10min\")",
+				QueryPatt:  "movingSum(second,\"10min\")",
+				QueryFrom:  10 + offset10mins,
+				QueryTo:    80,
+				Datapoints: getCopy(sumALast10mins),
+			},
+		},
+		offset10mins,
+		"10min",
+		"sum",
+		1,
+		t)
+}
+
+func TestMovingWindowWithXFilesFactorFilter(t *testing.T) {
+	offset20mins := uint32(20)
+	testMovingWindow(
+		"movingSum with xFilesFactor 1/3 nulls in window",
+		[]models.Series{
+			{
+				Target:     "first",
+				QueryPatt:  "first",
+				QueryFrom:  10 + offset20mins,
+				QueryTo:    80,
+				Datapoints: getCopy(seriesB),
+			},
+		},
+		[]models.Series{
+			{
+				Target:     "movingSum(first,\"20min\")",
+				QueryPatt:  "movingSum(first,\"20min\")",
+				QueryFrom:  10 + offset20mins,
+				QueryTo:    80,
+				Datapoints: getCopy(sumBLast20minsWithOneThirdNullsInWindowAllowed),
+			},
+		},
+		offset20mins,
+		"20min",
+		"sum",
+		0.66, // xFilesFactor 2/3 non null
+		t)
+
+	testMovingWindow(
+		"movingSum with xFilesFactor 2/3 nulls in window",
+		[]models.Series{
+			{
+				Target:     "first",
+				QueryPatt:  "first",
+				QueryFrom:  10 + offset20mins,
+				QueryTo:    80,
+				Datapoints: getCopy(seriesB),
+			},
+		},
+		[]models.Series{
+			{
+				Target:     "movingSum(first,\"20min\")",
+				QueryPatt:  "movingSum(first,\"20min\")",
+				QueryFrom:  10 + offset20mins,
+				QueryTo:    80,
+				Datapoints: getCopy(sumBLast20minsWithTwoThirdNullsInWindowAllowed),
+			},
+		},
+		offset20mins,
+		"20min",
+		"sum",
+		0.33, // xFilesFactor 1/3 non null
+		t)
+}
+
+func testMovingWindow(name string, in []models.Series, out []models.Series, offset uint32, windowSize, fn string, xFilesFactor float64, t *testing.T) {
+
+	// Below we simulate the effect of the time
+	// modifications in the function Context
+	// QueryFrom is reset. QueryTo is left untouched.
```

Comment:
Better to actually call Context. See the [timeshift test](https://github.com/bloomberg/metrictank/pull/89/files#diff-8a4e0f0f11eec43dba88d419b0a57dcfR161)